src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+      otherInfo.classInfo = std::nullopt;
+      otherInfo.permutation = groupInfo.permutation;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 2A: This shape is already found in an existing nontrivial
+    // equivalence class. Join the class and try the next permutation.
+    if (groups[otherRep].classInfo) {
+      auto& classInfo = *groups[otherRep].classInfo;
+
+      [[maybe_unused]] size_t unionRep =
+        equivalenceClasses.getUnion(otherRep, groupRep);
+      assert(otherRep == unionRep);
+
+      // Since we matched shapes with `other`, unapplying its permutation gets
+      // us back to the canonical shape, from which we can apply a fresh
+      // permutation.
+      groupInfo.classInfo = std::nullopt;
+      groupInfo.permutation = otherInfo.permutation;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 2B: We need to join two matching unaffiliated groups into a new
+    // nontrivial equivalence class.
+    assert(!groups[groupRep].classInfo && !groups[otherRep].classInfo);
+    assert(group == groupRep && other == otherRep);
+
+    // We are canonicalizing and reinserting the shape for other, so remove it
+    // from the map under its current shape.
+    groupShapeIndices.erase(it);
+
+    // Put the types for both groups in to the canonical order.
+    auto permutation = getCanonicalPermutation(groupInfo.group);
+    groupInfo.permutation = otherInfo.permutation = std::move(permutation);
+
+    // Set up `other` to be the representative element of the equivalence class.
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
```

Comment:
Is "does not change the graph" used in the meaning of "is isomorphic" here? If not, what is the difference?

Reply:
Any permutation of the types is an isomorphism, so in general an isomorphism of a rec group does not have the same rec group shape. An automorphism is a shape-preserving isomorphism.

Reply:
Then does "does not change the graph" mean "is automorphic"? I'm just trying to precisely understand what formal term those informal words mean.

Reply:
In other words, "does not change" is not clear to me. Is it not changing the graph module some property, or not at all in some absolute sense, etc.

I suspect examples might be very useful in a bunch of places here.

Reply:
By "does not change the graph," I mean "does not change the _shape_ of the graph," so yes, an isomorphism that does not change the (shape of the) graph is an automorphism.

Reply:
I am truly sorry, but I'm still stuck here. I think there are just some assumptions in your terms that I am not familiar with or have not internalized.

Specifically, when you say "does not change the *shape* of the graph", then I'm not sure if I should be imagining "shape" in the sense of rec group canonicalization or some other sense. Does the shape include the order, the indexes, etc. - ?

I think concrete examples could quickly solve all of this. Here is the type of comment that I think would be great at the top of this function, that would make all of this extremely concrete:
```
Before proving anything, let us lay out the notation we'll be using. Consider
this example graph:

0: struct(0, 1)
1: struct(1, 0)
2: struct(0, 1)

(Here the first line means that type 0 is a struct with fields of type 0 and 1.)
Note that any permutation of the vertices in this graph is an isomorphism, for
example, the permutation [2,1,0] will produce

0: struct(2, 1)
1: struct(1, 2)
2: struct(2, 1)

This is isomorphic to the previous graph even though the indexes are different,
because [..explanation..]. And it is/isn't an automorphism because [..reasons..].
Here is an example that is the reverse (isn't/is an automorphism) [..example..].
```

Reply:
Yes, by "shape" I always mean the property the rec group canonicalization algorithm is comparing across rec groups to see if they are equivalent.

I'll add examples to the comments, but the comments are going to become very very long 🥲

Reply:
If it's a hassle, no worries - that it is the rec group meaning answers my direct question here.